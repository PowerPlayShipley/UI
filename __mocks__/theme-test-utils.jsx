// TODO: Fix this??

import * as React from 'react'
import userEvent from '@testing-library/user-event'
import { render as rtlRender } from '@testing-library/react'

import { ThemeProvider } from "styled-components";

import theme from '../src/themes'
import GlobalStyle from "../src/components/GlobalStyle";

function renderWithTheme(ui, { ...options } = {}) {
  const Wrapper = ({ children }) => (
    <ThemeProvider theme={theme}>
      <GlobalStyle />
      {children}
      <div id='modal-root' />
    </ThemeProvider>
  )

  return rtlRender(ui, { wrapper: Wrapper, ...options })
}

// THE BELLOW COME FROM `https://github.com/styled-components/styled-components/blob/0aa3170c255a49cd41c3fbeb2b8051b5d132f229/src/test/utils.js`

const stripComments = (str) => str.replace(/\/\*.*?\*\/\n?/g, '');

const stripWhitespace = (str) => str
  .trim()
  .replace(/([;\{\}])/g, '$1  ')
  .replace(/\s+/g, ' ');

const getCSS = (scope) =>
  Array.from(scope.querySelectorAll('style'))
    .map(tag => tag.innerHTML)
    .join('\n')
    .replace(/ {/g, '{')
    .replace(/:\s+/g, ':')
    .replace(/:\s+;/g, ':;');

function expectCSSMatches ( _expectation, opts = { ignoreWhitespace: true }) {
  // NOTE: This should normalise both CSS strings to make irrelevant mismatches less likely
  const expectation = _expectation
    .replace(/ {/g, '{')
    .replace(/:\s+/g, ':')
    .replace(/:\s+;/g, ':;');

  const css = getCSS(document);

  if (opts.ignoreWhitespace) {
    const stripped = stripWhitespace(stripComments(css));
    expect(stripped).toEqual(stripWhitespace(expectation));
    return stripped;
  } else {
    expect(css).toEqual(expectation);
    return css;
  }
};

export * from '@testing-library/react';

export { renderWithTheme, expectCSSMatches };
