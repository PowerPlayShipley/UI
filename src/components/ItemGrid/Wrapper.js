import styled from "styled-components";

const Wrapper = styled.div`
  display: flex;
  flex: 1;
  
  // Very crude but will work
  div[disabled],  div.disabled, div[aria-disabled=true]  {
    pointer-events: none;
    
    color: ${({ theme }) => theme.main.colors.text.disabled };
    opacity: 65%;
  }
`

Wrapper.displayName = 'Grid'

export default Wrapper
