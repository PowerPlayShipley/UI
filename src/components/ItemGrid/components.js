import styled from "styled-components";

export const Icon = styled.div`
  display: flex;
  align-items: center;
  
  height: 24px;
  width: 24px;
  
  justify-content: center;
`

export const Item = styled.div`
  display: flex;
  align-items: center;
  
  border-radius: 3px;
  padding: 4px;
  
  color: #222f3e;
  
  cursor: pointer;
  user-select: none;
  
  &:hover {
    background-color: #dee0e2;
  }
`

export const Grid = styled.div`
  display: flex;
  flex-wrap: wrap;
  
  max-height: 200px;
  
  overflow-x: hidden;
  overflow-y: auto;
  
  padding: 0;
`
