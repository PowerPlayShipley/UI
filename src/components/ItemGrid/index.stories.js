import ItemGrid from "./index";

export default {
  title: 'ItemGrid',
  component: ItemGrid
}

const Template = ({ items, length, ...rest }) => {
  const arr = Array.from({ length }, () => items[getRandomInt(items.length)])
  return (<ItemGrid items={arr} {...rest} />);
}

export const Default = Template.bind({});
Default.args = {
  length: 100,
  items: [ "😴", "😄", "😃", "⛔", "🎠", "🚓", "🚇"]
}

function getRandomInt(max) {
  return Math.floor(Math.random() * max);
}
