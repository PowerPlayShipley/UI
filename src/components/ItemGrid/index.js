import React, { useCallback } from 'react'
import PropTypes from 'prop-types'

import Wrapper from "./Wrapper";
import { Icon, Item, Grid } from './components'

const ItemGrid = ({ items, onClick }) => {
  const handleClick = useCallback((e, item, idx) => {
    e.preventDefault()
    onClick && onClick(item, idx)
  }, [onClick])

  return (
    <Wrapper>
      <Grid>
        {items.map((item, idx) => (
          <Item data-item={item} key={idx} onClick={(e) => handleClick(e, item, idx)}>
            <Icon>{item}</Icon>
          </Item>
        ))}
      </Grid>
    </Wrapper>
  )
}

ItemGrid.propTypes = {
  items: PropTypes.array.isRequired,
  onClick: PropTypes.func
}

export default ItemGrid
