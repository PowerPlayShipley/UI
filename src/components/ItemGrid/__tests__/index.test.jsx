/**
 * @jest-environment jsdom
 */

import React from 'react'
import { renderWithTheme } from '../../../../__mocks__/theme-test-utils'

import ItemGrid from "../index";

describe('Item Grid', () => {
  it('should create valid snapshot', function () {
    const { container } = renderWithTheme(<ItemGrid items={['1', '2', '3', '4']} />)
    expect(container.firstChild).toMatchSnapshot()
  });

  it('should load correct amount of items', function () {
    const items = ['1', '2', '3', '4']
    const { container, getByText } = renderWithTheme(<ItemGrid items={items} />)

    for (const item of items) {
      const child = getByText(item)
      expect(container).toContainElement(child)
    }
  });
})
