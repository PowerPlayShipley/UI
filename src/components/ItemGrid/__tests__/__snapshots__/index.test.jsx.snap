// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Item Grid should create valid snapshot 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 24px;
  width: 24px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 3px;
  padding: 4px;
  color: #222f3e;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c2:hover {
  background-color: #dee0e2;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  max-height: 200px;
  overflow-x: hidden;
  overflow-y: auto;
  padding: 0;
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <div
      class="c2"
      data-item="1"
    >
      <div
        class="c3"
      >
        1
      </div>
    </div>
    <div
      class="c2"
      data-item="2"
    >
      <div
        class="c3"
      >
        2
      </div>
    </div>
    <div
      class="c2"
      data-item="3"
    >
      <div
        class="c3"
      >
        3
      </div>
    </div>
    <div
      class="c2"
      data-item="4"
    >
      <div
        class="c3"
      >
        4
      </div>
    </div>
  </div>
</div>
`;
